.config/doom/config.org
#+TITLE: fqs's GNU Emacs Config
#+AUTHOR: firmanqs
#+DESCRIPTION: fqs's personal Emacs config.
#+STARTUP: showeverything
#+OPTIONS: toc:3

* TABLE OF CONTENTS :toc:
- [[#all-config-here-im-to-the-point-man][ALL CONFIG HERE I'M /TO THE POINT/ MAN]]
- [[#appearance][APPEARANCE]]
  - [[#fonts][Fonts]]
  - [[#modeline][Modeline]]
  - [[#theme][Theme]]
  - [[#other][Other]]
- [[#evil][EVIL]]
- [[#perspective][PERSPECTIVE]]
- [[#projectile][PROJECTILE]]
- [[#development][DEVELOPMENT]]
  - [[#for-php][for PHP]]
  - [[#for-latex][for latex]]
  - [[#for-lua][for Lua]]
  - [[#for-javascript][for javascript]]
- [[#send-deleted-file-to-the-trash-first][SEND DELETED FILE TO THE TRASH FIRST]]
- [[#org][ORG]]
- [[#custom-keybindings][Custom Keybindings]]

* ALL CONFIG HERE I'M /TO THE POINT/ MAN
#+begin_src emacs-lisp
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-
;; Place your private configuration here! Remember, you do not need to run 'doom
;; sync' after modifying this file!
;; Some functionality uses this to identify you, e.g. GPG configuration, email
;; clients, file templates and snippets. It is optional.
;; (setq user-full-name "John Doe"
;;       user-mail-address "john@doe.com")

;; Whenever you reconfigure a package, make sure to wrap your config in an
;; `after!' block, otherwise Doom's defaults may override your settings. E.g.
;;
;;   (after! PACKAGE
;;     (setq x y))
;;
;; The exceptions to this rule:
;;
;;   - Setting file/directory variables (like `org-directory')
;;   - Setting variables which explicitly tell you to set them before their
;;     package is loaded (see 'C-h v VARIABLE' to look up their documentation).
;;   - Setting doom variables (which start with 'doom-' or '+').
;;
;; Here are some additional functions/macros that will help you configure Doom.
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package!' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', relative to
;;   this file. Emacs searches the `load-path' when you load packages with
;;   `require' or `use-package'.
;; - `map!' for binding new keys
#+end_src

* APPEARANCE
** Fonts
#+begin_src emacs-lisp
;; Doom exposes five (optional) variables for controlling fonts in Doom:
;;
;; - `doom-font' -- the primary font to use
;; - `doom-variable-pitch-font' -- a non-monospace font (where applicable)
;; - `doom-big-font' -- used for `doom-big-font-mode'; use this for
;;   presentations or streaming.
;; - `doom-symbol-font' -- for symbols
;; - `doom-serif-font' -- for the `fixed-pitch-serif' face
;;
;; See 'C-h v doom-font' for documentation and more examples of what they
;; accept. For example:
;;
;;(setq doom-font (font-spec :family "Fira Code" :size 12 :weight 'semi-light)
;;      doom-variable-pitch-font (font-spec :family "Fira Sans" :size 13))
;;
;; If you or Emacs can't find your font, use 'M-x describe-font' to look them
;; up, `M-x eval-region' to execute elisp code, and 'M-x doom/reload-font' to
;; refresh your font settings. If Emacs still can't find your font, it likely
;; wasn't installed correctly. Font issues are rarely Doom issues!

;; This determines the style of line numbers in effect. If set to `nil', line
;; numbers are disabled. For relative line numbers, set this to `relative'.
(setq doom-font (font-spec :family "JetBrainsMono Nerd Font" :size 18)
      doom-variable-pitch-font (font-spec :family "JetBrainsMono Nerd Font" :size 18)
      doom-symbol-font (font-spec :family "JetBrainsMono Nerd Font" :size 18)
      doom-serif-font (font-spec :family "JetBrainsMono Nerd Font" :size 18)
      doom-big-font (font-spec :family "JetBrainsMono Nerd Font" :size 24))
#+end_src

** Modeline
#+begin_src emacs-lisp
(setq doom-modeline-buffer-file-name-style 'file-name
      doom-modeline-icon nil
      doom-modeline-lsp-icon t
      doom-modeline-buffer-encoding t
      doom-modeline-indent-info t
      doom-modeline-lsp t
      doom-modeline-modal nil)
#+end_src

** Theme
#+begin_src emacs-lisp
(add-to-list 'custom-theme-load-path "~/.dotfiles/.config/doom/themes/everforest")
(add-to-list 'custom-theme-load-path "~/.dotfiles/.config/doom/themes/doom-rose-pine")

;; There are two ways to load a theme. Both assume the theme is installed and
;; available. You can either set `doom-theme' or manually load a theme with the
;; `load-theme' function. This is the default:
;; (setq doom-theme 'gruber-darker)
;; (setq doom-theme 'doom-ir-black)
;; (setq doom-theme 'doom-tomorrow-night)
;; (setq doom-theme 'doom-solarized-dark-high-contrast)
;; (setq doom-theme 'doom-miramare)
;; (setq doom-theme 'doom-sourcerer)
;; (setq doom-theme 'nord)
(setq doom-theme 'doom-gruvbox)
;; (setq doom-theme 'everforest-hard-dark)
;; (setq doom-theme 'doom-material)
;; (setq catppuccin-flavor 'mocha)
#+end_src

** Other
*** Cursor
#+begin_src emacs-lisp
(setq evil-insert-state-cursor '(box "yellow"))
#+end_src
*** Line Number
#+begin_src emacs-lisp
;; (setq display-line-numbers-type 'relative) ;; relative line numbering for chad
(setq display-line-numbers-type 'visual) ;; relative line numbering for chad
#+end_src
*** Etc
#+begin_src emacs-lisp
(global-auto-revert-mode t)  ;; Automatically show changes if the file has changed
(setq scroll-margin 3)
(global-visual-line-mode t)
(column-number-mode 1) ;; show column where the cursor at right now in the statusline.
(setq resize-mini-windows nil)
#+end_src

* EVIL
#+begin_src emacs-lisp
(with-eval-after-load 'evil
  (define-key evil-motion-state-map (kbd "g C-a") #'evil-numbers/inc-at-pt-incremental)
  (define-key evil-motion-state-map (kbd "g C-x") #'evil-numbers/dec-at-pt-incremental))
#+end_src

* PERSPECTIVE
Multiple named workspace in Emacs. Config source [[https://gitlab.com/dwt1/dotfiles/-/blob/master/.config/doom/config.org?ref_type=heads][Derek Taylor config]]. (I disabled it because apparently doom emacs already has something called "Workspace" with the key SPC TAB and this is really amazing)
#+begin_src emacs-lisp
;; (map! :lrader
;;       :desc "Switch to perspective NAME"       "p s" #'persp-switch
;;       :desc "Switch to buffer in perspective"  "p b" #'persp-switch-to-buffer
;;       :desc "Switch to next perspective"       "p n" #'persp-next
;;       :desc "Switch to previous perspective"   "p p" #'persp-prev
;;       :desc "Add a buffer current perspective" "p a" #'persp-add-buffer
;;       :desc "Remove perspective by name"       "p r" #'persp-remove-by-name)
#+end_src

* PROJECTILE
#+begin_src emacs-lisp
(setq projectile-project-search-path '("~/Starship/"
                                       "~/.dotfiles"))
#+end_src

* DEVELOPMENT
** for PHP
#+begin_src emacs-lisp
(add-load-path! "~/.config/doom/extra-packages/php-extras")
(eval-after-load 'php-mode
  (require 'php-extras))
#+end_src
** for latex
#+begin_src emacs-lisp
;; working with multiple file
(setq TeX-auto-save t)
(setq TeX-parse-self t)
(setq-default TeX-master nil)

(setq LaTeX-indent-level 4)
(setq tex-indent-basic 4)
(setq TeX-brace-indent-level 4)
(setq LaTeX-item-indent -2)
;; for compiling with latexmk
(add-hook 'LaTeX-mode-hook
          (lambda () (local-set-key (kbd "M-s") #'TeX-command-run-all)))
#+end_src

** for Lua
#+begin_src emacs-lisp
(setq lua-indent-level 3)
#+end_src

** for javascript
#+begin_src emacs-lisp
(setq js-indent-level 4)
#+end_src

* SEND DELETED FILE TO THE TRASH FIRST
#+begin_src emacs-lisp
(setq delete-by-moving-to-trash t
      trash-directory "~/.local/share/Trash/files/")
#+end_src

* ORG
#+begin_src emacs-lisp
;; If you use `org' and don't want your org files in the default location below,
;; change `org-directory'. It must be set before org loads!
(setq org-directory "~/starship/myNote")

(after! org
  (setq org-default-notes-file (expand-file-name "notes.org" org-directory)
        org-ellipsis " 󱞣 "
        org-superstar-headline-bullets-list '("◉" "●" "○" "◆" "●" "○" "◆")
        ;; org-superstar-headline-bullets-list '("✽" "✾" "❆" "❆" "❁" "❅" "✼")
        ;; org-superstar-headline-bullets-list '("◐" "◑" "◒" "◓" "⚈" "⚉" "⊗")
        org-superstar-itembullet-alist '((?+ . ?➤) (?- . ?✦)) ; changes +/- symbols in item lists
        ;; org-superstar-itembullet-alist '((?+ . ?➤) (?- . ?❍)) ; changes +/- symbols in item lists
        org-log-done 'time
        org-hide-emphasis-markers t))
(setq org-auto-tangle-default t)

(use-package! org-transclusion
              :after org
              :init
              (map!
		:map global-map "<f12>" #'org-transclusion-add
		:leader
		:prefix "n"
		:desc "Org Transclusion Mode" "t" #'org-transclusion-mode))
#+end_src

* Custom Keybindings
#+begin_src emacs-lisp
(map! "M-j" #'drag-stuff-down
      "M-k" #'drag-stuff-up
      "M-l" #'drag-stuff-right
      "M-h" #'drag-stuff-left)

;; remaping unusefull evil bindings
(map! :nvi "C-e" #'evil-end-of-visual-line)
(map! :m "C-a" #'evil-next-line-1-first-non-blank)
#+end_src


.config/doom/config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;; Place your private configuration here! Remember, you do not need to run 'doom
;; sync' after modifying this file!

;; Some functionality uses this to identify you, e.g. GPG configuration, email
;; clients, file templates and snippets. It is optional.
;; (setq user-full-name "John Doe"
;;       user-mail-address "john@doe.com")

;; Doom exposes five (optional) variables for controlling fonts in Doom:
;;
;; - `doom-font' -- the primary font to use
;; - `doom-variable-pitch-font' -- a non-monospace font (where applicable)
;; - `doom-big-font' -- used for `doom-big-font-mode'; use this for
;;   presentations or streaming.
;; - `doom-symbol-font' -- for symbols
;; - `doom-serif-font' -- for the `fixed-pitch-serif' face
;;
;; See 'C-h v doom-font' for documentation and more examples of what they
;; accept. For example:
;;
;;(setq doom-font (font-spec :family "Fira Code" :size 12 :weight 'semi-light)
;;      doom-variable-pitch-font (font-spec :family "Fira Sans" :size 13))
;;
;; If you or Emacs can't find your font, use 'M-x describe-font' to look them
;; up, `M-x eval-region' to execute elisp code, and 'M-x doom/reload-font' to
;; refresh your font settings. If Emacs still can't find your font, it likely
;; wasn't installed correctly. Font issues are rarely Doom issues!

;; There are two ways to load a theme. Both assume the theme is installed and
;; available. You can either set `doom-theme' or manually load a theme with the
;; `load-theme' function. This is the default:
;; (setq doom-theme 'gruber-darker)
;; (setq doom-theme 'doom-ir-black)
;; (setq doom-theme 'doom-tomorrow-night)
;; (setq doom-theme 'doom-solarized-dark-high-contrast)
;; (setq doom-theme 'doom-miramare)
;; (setq doom-theme 'doom-sourcerer)
;; (setq doom-theme 'nord)
(setq doom-theme 'everforest-hard-dark)

;; This determines the style of line numbers in effect. If set to `nil', line
;; numbers are disabled. For relative line numbers, set this to `relative'.
(setq doom-font (font-spec :family "JetBrainsMono Nerd Font" :size 18)
      doom-variable-pitch-font (font-spec :family "JetBrainsMono Nerd Font" :size 18)
      doom-symbol-font (font-spec :family "JetBrainsMono Nerd Font" :size 18)
      doom-serif-font (font-spec :family "JetBrainsMono Nerd Font" :size 18)
      doom-big-font (font-spec :family "JetBrainsMono Nerd Font" :size 24))

(setq evil-insert-state-cursor '(box "yellow"))

;; (setq display-line-numbers-type 'relative) ;; relative line numbering for chad
(setq display-line-numbers-type 'visual) ;; relative line numbering for chad

(global-auto-revert-mode t)  ;; Automatically show changes if the file has changed
(setq scroll-margin 3)
(global-visual-line-mode t)
(column-number-mode 1) ;; show column where the cursor at right now in the statusline.

(with-eval-after-load 'evil
  (define-key evil-motion-state-map (kbd "g C-a") #'evil-numbers/inc-at-pt-incremental)
  (define-key evil-motion-state-map (kbd "g C-x") #'evil-numbers/dec-at-pt-incremental))

;; Whenever you reconfigure a package, make sure to wrap your config in an
;; `after!' block, otherwise Doom's defaults may override your settings. E.g.
;;
;;   (after! PACKAGE
;;     (setq x y))
;;
;; The exceptions to this rule:
;;
;;   - Setting file/directory variables (like `org-directory')
;;   - Setting variables which explicitly tell you to set them before their
;;     package is loaded (see 'C-h v VARIABLE' to look up their documentation).
;;   - Setting doom variables (which start with 'doom-' or '+').
;;
;; Here are some additional functions/macros that will help you configure Doom.
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package!' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', relative to
;;   this file. Emacs searches the `load-path' when you load packages with
;;   `require' or `use-package'.
;; - `map!' for binding new keys
;;
;; To get information about any of these functions/macros, move the cursor over
;; the highlighted symbol at press 'K' (non-evil users must press 'C-c c k').
;; This will open documentation for it, including demos of how they are used.
;; Alternatively, use `C-h o' to look up a symbol (functions, variables, faces,
;; etc).
;;
;; You can also try 'gd' (or 'C-c c d') to jump to their definition and see how
;; they are implemented.

;; (set-frame-parameter nil 'alpha-background 90)
;; (add-to-list 'default-frame-alist '(alpha-background . 90))

;; (map! :lrader
;;       :desc "Switch to perspective NAME"       "p s" #'persp-switch
;;       :desc "Switch to buffer in perspective"  "p b" #'persp-switch-to-buffer
;;       :desc "Switch to next perspective"       "p n" #'persp-next
;;       :desc "Switch to previous perspective"   "p p" #'persp-prev
;;       :desc "Add a buffer current perspective" "p a" #'persp-add-buffer
;;       :desc "Remove perspective by name"       "p r" #'persp-remove-by-name)

(setq projectile-project-search-path '("~/Starship/" "~/.dotfiles")
      projectile-require-project-root t
      projectile-enable-caching nil)

;; working with multiple file
(setq TeX-auto-save t)
(setq TeX-parse-self t)
(setq-default TeX-master nil)

(setq LaTeX-indent-level 4)
(setq tex-indent-basic 4)
(setq TeX-brace-indent-level 4)
(setq LaTeX-item-indent -2)
;; for compiling with latexmk
(add-hook 'LaTeX-mode-hook
          (lambda () (local-set-key (kbd "M-s") #'TeX-command-run-all)))

(setq js-indent-level 4)

(setq delete-by-moving-to-trash t
      trash-directory "~/.local/share/Trash/files/")

(setq doom-modeline-buffer-file-name-style 'file-name)
(setq doom-modeline-icon t)

;; Whether display the icon for `major-mode'. It respects option `doom-modeline-icon'.
(setq doom-modeline-major-mode-icon t)

;; Whether display the colorful icon for `major-mode'.
;; It respects `nerd-icons-color-icons'.
(setq doom-modeline-major-mode-color-icon t)

;; Whether display the icon for the buffer state. It respects option `doom-modeline-icon'.
(setq doom-modeline-buffer-state-icon t)

;; Whether display the modification icon for the buffer.
;; It respects option `doom-modeline-icon' and option `doom-modeline-buffer-state-icon'.
(setq doom-modeline-buffer-modification-icon t)

;; Whether display the lsp icon. It respects option `doom-modeline-icon'.
(setq doom-modeline-lsp-icon t)

;; Whether to use unicode as a fallback (instead of ASCII) when not using icons.
;; (setq doom-modeline-unicode-fallback nil)

;; Whether display the buffer encoding.
(setq doom-modeline-buffer-encoding t)

;; Whether display the indentation information.
(setq doom-modeline-indent-info t)

;; Whether display the `lsp' state. Non-nil to display in the mode-line.
(setq doom-modeline-lsp t)

;; (setq modus-themes-bold-constructs t
;;       modus-themes-italic-constructs t
;;       modus-themes-fringes 'subtle
;;       modus-themes-org-blocks 'tinted-background
;;       modus-themes-prompts '(bold intense)
;;       modus-themes-paren-match '(bold intense)
;;       modus-themes-headings '((1 . (rainbow 1.4))
;;                               (2 . (rainbow 1.3))
;;                               (3 . (rainbow 1.2))))

;; (setq doom-theme 'modus-vivendi)

;; If you use `org' and don't want your org files in the default location below,
;; change `org-directory'. It must be set before org loads!
(setq org-directory "~/Starship/myNote")

(after! org
  (setq org-default-notes-file (expand-file-name "notes.org" org-directory)
        org-ellipsis " 󱞣 "
        org-superstar-headline-bullets-list '("◉" "●" "○" "◆" "●" "○" "◆")
        ;; org-superstar-headline-bullets-list '("✽" "✾" "❆" "❆" "❁" "❅" "✼")
        ;; org-superstar-headline-bullets-list '("◐" "◑" "◒" "◓" "⚈" "⚉" "⊗")
        org-superstar-itembullet-alist '((?+ . ?➤) (?- . ?✦)) ; changes +/- symbols in item lists
        ;; org-superstar-itembullet-alist '((?+ . ?➤) (?- . ?❍)) ; changes +/- symbols in item lists
        org-log-done 'time
        org-hide-emphasis-markers t))
(setq org-auto-tangle-default t)

(map! "M-j" #'drag-stuff-down
      "M-k" #'drag-stuff-up
      "M-l" #'drag-stuff-right
      "M-h" #'drag-stuff-left)

;; remaping unusefull evil bindings
(map! :nvi "C-e" #'evil-end-of-visual-line)
(map! :m "C-a" #'evil-next-line-1-first-non-blank)

(add-to-list 'custom-theme-load-path "~/.dotfiles/.config/doom/themes/everforest")
(add-to-list 'custom-theme-load-path "~/.dotfiles/.config/doom/themes/doom-rose-pine")
